File: src/mvmctl/models/host.py
"""Host data models."""

from __future__ import annotations

from dataclasses import dataclass, field

from mvmctl.utils.common import CommonUtils


@dataclass
class HostStateItem:
    """Host state record — maps to host_state table (singleton id=1)."""

    id: int
    initialized: bool
    mvm_group_created: bool
    sudoers_configured: bool
    default_network_created: bool
    initialized_at: str
    updated_at: str

    # Capacity detection fields (populated by host info --refresh or host init)
    hostname: str | None = None
    cpu_model: str | None = None
    cpu_vendor: str | None = None
    cpu_cores: int | None = None
    cpu_architecture: str | None = None
    numa_nodes: int | None = None
    memory_total_mib: int | None = None
    storage_total_bytes: int | None = None
    kernel_version: str | None = None
    os_release: str | None = None
    pid_max: int | None = None
    fd_max: int | None = None
    conntrack_max: int | None = None
    tap_devices_max: int | None = None
    ip_local_port_range: str | None = None
    detected_at: str | None = None

    # Virtualization detection fields
    cpu_has_vmx: int | None = None
    cpu_hypervisor: int | None = None
    nested_virt_available: int | None = None
    ept_available: int | None = None
    hugepage_count_2mb: int | None = None
    ksm_disabled: int | None = None
    cgroup_version: int | None = None
    swap_total_mib: int | None = None
    kernel_minimum_met: int | None = None

    def __post_init__(self) -> None:
        """Coerce bool fields loaded from SQLite."""
        CommonUtils.coerce_bool_fields(
            self,
            {
                "initialized",
                "mvm_group_created",
                "sudoers_configured",
                "default_network_created",
            },
        )


@dataclass
class HostStateChangeItem:
    """Host state change record — maps to host_state_changes table."""

    session_id: str
    init_timestamp: str
    setting: str
    mechanism: str
    applied_value: str
    reverted: bool
    change_order: int
    created_at: str

    id: int | None = None
    original_value: str | None = None
    reverted_at: str | None = None
    revert_mechanism: str | None = None

    def __post_init__(self) -> None:
        """Coerce bool fields loaded from SQLite."""
        CommonUtils.coerce_bool_fields(self, {"reverted"})


@dataclass
class HostHardware:
    """Detected host hardware capabilities."""

    hostname: str
    cpu_model: str
    cpu_vendor: str
    cpu_cores: int
    cpu_architecture: str
    numa_nodes: int
    memory_total_mib: int
    storage_total_bytes: int
    kernel_version: str
    os_release: str
    cpu_has_vmx: bool = False
    cpu_hypervisor: bool = False

    @classmethod
    def from_state(cls, state: HostStateItem) -> HostHardware | None:
        """Reconstruct HostHardware from stored state, or None if not yet detected."""
        if state.cpu_model is None:
            return None
        return cls(
            hostname=state.hostname or "",
            cpu_model=state.cpu_model or "",
            cpu_vendor=state.cpu_vendor or "",
            cpu_cores=state.cpu_cores or 0,
            cpu_architecture=state.cpu_architecture or "",
            numa_nodes=state.numa_nodes or 1,
            memory_total_mib=state.memory_total_mib or 0,
            storage_total_bytes=state.storage_total_bytes or 0,
            kernel_version=state.kernel_version or "",
            os_release=state.os_release or "",
            cpu_has_vmx=bool(state.cpu_has_vmx)
            if state.cpu_has_vmx is not None
            else False,
            cpu_hypervisor=bool(state.cpu_hypervisor)
            if state.cpu_hypervisor is not None
            else False,
        )


@dataclass
class HostLimits:
    """Detected host kernel limits."""

    pid_max: int
    fd_max: int
    conntrack_max: int
    tap_devices_max: int
    ip_local_port_range: tuple[int, int]
    nested_virt_available: bool = False
    ept_available: bool = False
    hugepage_count_2mb: int = 0
    ksm_disabled: bool = True
    cgroup_version: int = 1
    swap_total_mib: int = 0
    kernel_minimum_met: bool = False

    @classmethod
    def from_state(cls, state: HostStateItem) -> HostLimits | None:
        """Reconstruct HostLimits from stored state, or None if not yet detected."""
        if state.pid_max is None:
            return None
        port_range = (32768, 60999)
        if state.ip_local_port_range:
            try:
                parts = state.ip_local_port_range.split(",")
                if len(parts) == 2:
                    port_range = (int(parts[0]), int(parts[1]))
            except (ValueError, TypeError):
                pass
        return cls(
            pid_max=state.pid_max or 0,
            fd_max=state.fd_max or 0,
            conntrack_max=state.conntrack_max or 0,
            tap_devices_max=state.tap_devices_max
            if state.tap_devices_max is not None
            else 0,
            ip_local_port_range=port_range,
            nested_virt_available=bool(state.nested_virt_available)
            if state.nested_virt_available is not None
            else False,
            ept_available=bool(state.ept_available)
            if state.ept_available is not None
            else False,
            hugepage_count_2mb=state.hugepage_count_2mb or 0,
            ksm_disabled=bool(state.ksm_disabled)
            if state.ksm_disabled is not None
            else True,
            cgroup_version=state.cgroup_version or 1,
            swap_total_mib=state.swap_total_mib or 0,
            kernel_minimum_met=bool(state.kernel_minimum_met)
            if state.kernel_minimum_met is not None
            else False,
        )


@dataclass
class HostResources:
    """Current host resource usage and capacity projection."""

    memory_available_mib: int
    tap_devices_used: int
    pids_current: int
    fd_current: int
    conntrack_current: int
    arp_current: int
    storage_free_bytes: int
    recommended_max_vms: int
    limiting_resource: str | None
    modules_loaded: dict[str, bool] = field(default_factory=dict)
    swap_used_mib: int = 0
    hugepages_free_2mb: int = 0
    smt_active: bool = False
    nftables_available: bool = False
    iptables_available: bool = False
    cloud_localds_available: bool = False
    dev_kvm_status: str = ""
    user_in_kvm_group: bool = False
    dev_net_tun_accessible: bool = False


@dataclass
class ProbeCheck:
    """Result of a single pre-flight probe check."""

    name: str
    passed: bool
    message: str
    details: str | None = None


@dataclass
class ProbeResult:
    """Aggregated pre-flight probe results."""

    critical: list[ProbeCheck] = field(default_factory=list)
    warnings: list[ProbeCheck] = field(default_factory=list)
    info: list[ProbeCheck] = field(default_factory=list)

    @property
    def has_critical(self) -> bool:
        """Return True if there are any failed critical checks."""
        return bool(self.critical)


@dataclass
class HostInfo:
    """Aggregated host detection results — wraps all four host models into one."""

    state: HostStateItem
    resources: HostResources
    limits: HostLimits
    hardware: HostHardware

    def to_dict(self) -> dict[str, object]:
        """Build the standardised info response dict from all four sub-models."""
        return {
            "detected_at": self.state.detected_at or "",
            "hostname": self.hardware.hostname,
            "os": {
                "kernel": self.hardware.kernel_version,
                "release": self.hardware.os_release,
            },
            "cpu": {
                "model": self.hardware.cpu_model,
                "vendor": self.hardware.cpu_vendor,
                "cores": self.hardware.cpu_cores,
                "architecture": self.hardware.cpu_architecture,
                "numa_nodes": self.hardware.numa_nodes,
            },
            "virtualization": {
                "cpu_has_vmx": self.hardware.cpu_has_vmx,
                "nested_virt_available": self.limits.nested_virt_available,
                "ept_available": self.limits.ept_available,
                "hypervisor": self.hardware.cpu_hypervisor,
                "smt_active": self.resources.smt_active,
                "modules": dict(self.resources.modules_loaded),
            },
            "hugepages": {
                "count_2mb": self.limits.hugepage_count_2mb,
                "free_2mb": self.resources.hugepages_free_2mb,
            },
            "dependencies": {
                "nftables_available": self.resources.nftables_available,
                "iptables_available": self.resources.iptables_available,
                "cloud_localds_available": self.resources.cloud_localds_available,
                "dev_net_tun": self.resources.dev_net_tun_accessible,
            },
            "system": {
                "cgroup_version": self.limits.cgroup_version,
                "ksm_disabled": self.limits.ksm_disabled,
                "dev_kvm_status": self.resources.dev_kvm_status,
                "user_in_kvm_group": self.resources.user_in_kvm_group,
            },
            "memory": {
                "total_mib": self.hardware.memory_total_mib,
                "available_mib": self.resources.memory_available_mib,
                "swap_total_mib": self.limits.swap_total_mib,
                "swap_used_mib": self.resources.swap_used_mib,
            },
            "storage": {
                "total_bytes": self.hardware.storage_total_bytes,
                "free_bytes": self.resources.storage_free_bytes,
            },
            "kernel": {
                "version": self.hardware.kernel_version,
                "minimum_version_met": self.limits.kernel_minimum_met,
            },
            "limits": {
                "pid_max": self.limits.pid_max,
                "fd_max": self.limits.fd_max,
                "conntrack_max": self.limits.conntrack_max,
                "tap_devices_max": self.limits.tap_devices_max,
                "ip_local_port_range": list(self.limits.ip_local_port_range),
            },
            "capacity": {
                "current": {
                    "pids": self.resources.pids_current,
                    "fds": self.resources.fd_current,
                    "conntrack": self.resources.conntrack_current,
                    "tap_devices": self.resources.tap_devices_used,
                    "arp_entries": self.resources.arp_current,
                },
                "recommended_max_vms": self.resources.recommended_max_vms,
                "limiting_resource": self.resources.limiting_resource,
            },
            "setup": {
                "initialized": bool(self.state.initialized),
                "initialized_at": self.state.initialized_at,
            },
        }

File: internal/infra/model/host.go
package model

// ── HostStateItem ──

// HostStateItem matches Python's HostStateItem — singleton row (id=1).
type HostStateItem struct {
	ID                    int    `json:"id"                      db:"id"`
	Initialized           bool   `json:"initialized"             db:"initialized"`
	MvmGroupCreated       bool   `json:"mvm_group_created"       db:"mvm_group_created"`
	SudoersConfigured     bool   `json:"sudoers_configured"      db:"sudoers_configured"`
	DefaultNetworkCreated bool   `json:"default_network_created" db:"default_network_created"`
	InitializedAt         string `json:"initialized_at"          db:"initialized_at"`
	UpdatedAt             string `json:"updated_at"              db:"updated_at"`

	// Capacity detection fields
	Hostname          *string `json:"hostname,omitempty"`
	CPUModel          *string `json:"cpu_model,omitempty"`
	CPUVendor         *string `json:"cpu_vendor,omitempty"`
	CPUCores          *int    `json:"cpu_cores,omitempty"`
	CPUArchitecture   *string `json:"cpu_architecture,omitempty"`
	NumaNodes         *int    `json:"numa_nodes,omitempty"`
	MemoryTotalMiB    *int    `json:"memory_total_mib,omitempty"`
	StorageTotalBytes *int    `json:"storage_total_bytes,omitempty"`
	KernelVersion     *string `json:"kernel_version,omitempty"`
	OSRelease         *string `json:"os_release,omitempty"`
	PIDMax            *int    `json:"pid_max,omitempty"`
	FDMax             *int    `json:"fd_max,omitempty"`
	ConntrackMax      *int    `json:"conntrack_max,omitempty"`
	TAPDevicesMax     *int    `json:"tap_devices_max,omitempty"`
	IPLocalPortRange  *string `json:"ip_local_port_range,omitempty"`
	DetectedAt        *string `json:"detected_at,omitempty"`

	// Virtualization detection fields
	CPUHasVMX           *int `json:"cpu_has_vmx,omitempty"`
	CPUHypervisor       *int `json:"cpu_hypervisor,omitempty"`
	NestedVirtAvailable *int `json:"nested_virt_available,omitempty"`
	EPTAvailable        *int `json:"ept_available,omitempty"`
	HugepageCount2MB    *int `json:"hugepage_count_2mb,omitempty"`
	KSMDisabled         *int `json:"ksm_disabled,omitempty"`
	CgroupVersion       *int `json:"cgroup_version,omitempty"`
	SwapTotalMiB        *int `json:"swap_total_mib,omitempty"`
	KernelMinimumMet    *int `json:"kernel_minimum_met,omitempty"`
}

// ── HostStateChangeItem ──

// HostStateChangeItem matches Python's HostStateChangeItem.
type HostStateChangeItem struct {
	SessionID       string  `json:"session_id"`
	InitTimestamp   string  `json:"init_timestamp"`
	Setting         string  `json:"setting"`
	Mechanism       string  `json:"mechanism"`
	AppliedValue    string  `json:"applied_value"`
	Reverted        bool    `json:"reverted"`
	ChangeOrder     int     `json:"change_order"`
	CreatedAt       string  `json:"created_at"`
	ID              *int    `json:"id,omitempty"`
	OriginalValue   *string `json:"original_value,omitempty"`
	RevertedAt      *string `json:"reverted_at,omitempty"`
	RevertMechanism *string `json:"revert_mechanism,omitempty"`
}

// ── HostHardware ──

// HostHardware matches Python's HostHardware.
type HostHardware struct {
	Hostname          string `json:"hostname"`
	CPUModel          string `json:"cpu_model"`
	CPUVendor         string `json:"cpu_vendor"`
	CPUCores          int    `json:"cpu_cores"`
	CPUArchitecture   string `json:"cpu_architecture"`
	NumaNodes         int    `json:"numa_nodes"`
	MemoryTotalMiB    int    `json:"memory_total_mib"`
	StorageTotalBytes int    `json:"storage_total_bytes"`
	KernelVersion     string `json:"kernel_version"`
	OSRelease         string `json:"os_release"`
	CPUHasVMX         bool   `json:"cpu_has_vmx"`
	CPUHypervisor     bool   `json:"cpu_hypervisor"`
}

// ── HostLimits ──

// HostLimits matches Python's HostLimits.
type HostLimits struct {
	PIDMax              int    `json:"pid_max"`
	FDMax               int    `json:"fd_max"`
	ConntrackMax        int    `json:"conntrack_max"`
	TAPDevicesMax       int    `json:"tap_devices_max"`
	IPLocalPortRange    [2]int `json:"ip_local_port_range"`
	NestedVirtAvailable bool   `json:"nested_virt_available"`
	EPTAvailable        bool   `json:"ept_available"`
	HugepageCount2MB    int    `json:"hugepage_count_2mb"`
	KSMDisabled         bool   `json:"ksm_disabled"`
	CgroupVersion       int    `json:"cgroup_version"`
	SwapTotalMiB        int    `json:"swap_total_mib"`
	KernelMinimumMet    bool   `json:"kernel_minimum_met"`
}

// ── HostResources ──

// HostResources matches Python's HostResources.
type HostResources struct {
	MemoryAvailableMiB    int             `json:"memory_available_mib"`
	TAPDevicesUsed        int             `json:"tap_devices_used"`
	PIDsCurrent           int             `json:"pids_current"`
	FDCurrent             int             `json:"fd_current"`
	ConntrackCurrent      int             `json:"conntrack_current"`
	ARPCurrent            int             `json:"arp_current"`
	StorageFreeBytes      int             `json:"storage_free_bytes"`
	RecommendedMaxVMs     int             `json:"recommended_max_vms"`
	LimitingResource      *string         `json:"limiting_resource,omitempty"`
	ModulesLoaded         map[string]bool `json:"modules_loaded"`
	SwapUsedMiB           int             `json:"swap_used_mib"`
	HugepagesFree2MB      int             `json:"hugepages_free_2mb"`
	SMTActive             bool            `json:"smt_active"`
	NftablesAvailable     bool            `json:"nftables_available"`
	IptablesAvailable     bool            `json:"iptables_available"`
	CloudLocaldsAvailable bool            `json:"cloud_localds_available"`
	DevKVMStatus          string          `json:"dev_kvm_status"`
	UserInKVMGroup        bool            `json:"user_in_kvm_group"`
	DevNetTUNAccessible   bool            `json:"dev_net_tun_accessible"`
}

// ── ProbeCheck ──

// ProbeCheck matches Python's ProbeCheck.
type ProbeCheck struct {
	Name    string `json:"name"`
	Passed  bool   `json:"passed"`
	Message string `json:"message"`
	Details string `json:"details,omitempty"`
}

// ── ProbeResult ──

// ProbeResult matches Python's ProbeResult.
type ProbeResult struct {
	Critical []ProbeCheck `json:"critical"`
	Warnings []ProbeCheck `json:"warnings"`
	Info     []ProbeCheck `json:"info"`
}

// ── HostInfo ──

// HostInfo matches Python's HostInfo — aggregated host detection results.
type HostInfo struct {
	State     HostStateItem `json:"state"`
	Resources HostResources `json:"resources"`
	Limits    HostLimits    `json:"limits"`
	Hardware  HostHardware  `json:"hardware"`
}
